File: py/src/skkuverse_crawler/notices/strategies/gnuboard_custom.py
from __future__ import annotations

import re
from urllib.parse import urlparse

from ...shared.fetcher import Fetcher
from ...shared.logger import get_logger
from ..models import NoticeDetail, NoticeListItem
from ..parser import load_html, extract_text, extract_attr
from ..types import DetailRef

logger = get_logger("gnuboard_custom")


def clean_hwp_artifacts(html: str) -> str:
    cleaned = re.sub(r"<!--\[data-hwpjson\][\s\S]*?\[data-hwpjson\]-->", "", html)
    cleaned = re.sub(r'\s*data-hwpjson="[^"]*"', "", cleaned)
    return cleaned


class GnuboardCustomStrategy:
    def __init__(self, fetcher: Fetcher) -> None:
        self.fetcher = fetcher

    async def crawl_list(self, config: dict, page: int) -> list[NoticeListItem]:
        page_num = page + 1
        url = f"{config['baseUrl']}?{config['boardParam']}={config['boardName']}&page={page_num}"

        logger.info("fetching_gnuboard_custom_list", url=url, page=page)
        html = await self.fetcher.fetch(url)
        soup = load_html(html)

        items: list[NoticeListItem] = []
        selectors = config["selectors"]

        for tr in soup.select(selectors["listRow"]):
            try:
                tds = tr.select("td")
                if len(tds) < 2:
                    continue

                # Skip pinned on page > 0
                if page > 0 and tr.select_one('img[src*="btn_notice"]'):
                    continue

                title_link = tr.select_one(selectors["titleLink"])
                title = extract_text(title_link).strip()
                if not title:
                    continue

                href = extract_attr(title_link, "href") or ""
                m = re.search(r"num=(\d+)", href)
                if not m:
                    continue
                article_no = int(m.group(1))

                date = extract_text(tr.select_one(selectors["date"]))

                meta_text = extract_text(tr.select_one(selectors["meta"]))
                parts = [s.strip() for s in meta_text.split("|")]
                author = parts[0] if parts else ""
                vm = re.search(r"조회수\s*:\s*(\d+)", meta_text)
                views = int(vm.group(1)) if vm else 0

                items.append(NoticeListItem(
                    articleNo=article_no, title=title, category="",
                    author=author, date=date, views=views, detailPath=href,
                ))
            except Exception as exc:
                logger.warning("parse_gnuboard_custom_failed", error=str(exc))

        logger.info("parsed_gnuboard_custom_list", page=page, count=len(items))
        return items

    async def crawl_detail(self, ref: DetailRef, config: dict) -> NoticeDetail | None:
        if ref["detailPath"].startswith("http"):
            url = ref["detailPath"]
        elif ref["detailPath"].startswith("?"):
            url = f"{config['baseUrl']}{ref['detailPath']}"
        else:
            url = f"{config['baseUrl']}?{config['boardParam']}={config['boardName']}&mode={config['detailMode']}&num={ref['articleNo']}"

        try:
            html = await self.fetcher.fetch(url)
            soup = load_html(html)
            selectors = config["selectors"]

            content_el = soup.select_one(selectors["detailContent"])
            raw_html = content_el.decode_contents().strip() if content_el else ""
            content = clean_hwp_artifacts(raw_html)
            content_soup = load_html(content)
            content_text = content_soup.get_text(strip=True)

            attachments: list[dict[str, str]] = []
            parsed = urlparse(config["baseUrl"])
            origin = f"{parsed.scheme}://{parsed.netloc}"

            for a in soup.select(selectors["detailAttachment"]):
                name = a.get_text(strip=True)
                file_href = extract_attr(a, "href") or ""
                if not name or not file_href:
                    continue
                if file_href.startswith("/"):
                    file_href = f"{origin}{file_href}"
                elif not file_href.startswith("http"):
                    file_href = f"{origin}/{file_href}"
                attachments.append({"name": name, "url": file_href})

            return NoticeDetail(content=content, contentText=content_text, attachments=attachments)
        except Exception as exc:
            logger.error("gnuboard_custom_detail_failed", articleNo=ref["articleNo"], error=str(exc))
            return None

File: src/notices/strategies/gnuboard-custom.ts
import { Fetcher } from '../../shared/fetcher.js';
import { loadHtml } from '../parser.js';
import { NoticeListItem, NoticeDetail } from '../normalizer.js';
import type { GnuboardCustomDepartmentConfig, CrawlStrategy, DetailRef } from '../types.js';
import logger from '../../shared/logger.js';

/**
 * Clean HWP editor artifacts from HTML content.
 * Removes <!--[data-hwpjson]...[data-hwpjson]--> comments and data-hwpjson attributes.
 */
function cleanHwpArtifacts(html: string): string {
  // Remove HWP JSON comments: <!--[data-hwpjson]...[data-hwpjson]-->
  let cleaned = html.replace(/<!--\[data-hwpjson\][\s\S]*?\[data-hwpjson\]-->/g, '');
  // Remove data-hwpjson attributes from tags
  cleaned = cleaned.replace(/\s*data-hwpjson="[^"]*"/g, '');
  return cleaned;
}

export class GnuboardCustomStrategy implements CrawlStrategy {
  private fetcher: Fetcher;

  constructor(fetcher: Fetcher) {
    this.fetcher = fetcher;
  }

  async crawlList(config: GnuboardCustomDepartmentConfig, page: number): Promise<NoticeListItem[]> {
    const pageNum = page + 1; // 0-based → 1-based
    const url = `${config.baseUrl}?${config.boardParam}=${config.boardName}&page=${pageNum}`;

    logger.info({ url, page }, 'Fetching gnuboard-custom list page');
    const html = await this.fetcher.fetch(url);
    const $ = loadHtml(html);

    const items: NoticeListItem[] = [];

    $(config.selectors.listRow).each((_i, el) => {
      try {
        const $tr = $(el);
        const $tds = $tr.find('td');
        if ($tds.length < 2) return; // skip header rows

        // Skip notice rows (pinned) on page > 0 to avoid duplicates
        if (page > 0 && $tr.find('img[src*="btn_notice"]').length > 0) return;

        // Title from configured selector
        const $titleLink = $tr.find(config.selectors.titleLink);
        const title = $titleLink.text().trim();
        if (!title) return;

        const href = $titleLink.attr('href') || '';

        // Extract articleNo via num=(\d+)
        const articleNoMatch = href.match(/num=(\d+)/);
        if (!articleNoMatch) {
          logger.warn({ href, title }, 'Could not extract num from gnuboard-custom href');
          return;
        }
        const articleNo = parseInt(articleNoMatch[1], 10);

        // Date from configured selector
        const date = $tr.find(config.selectors.date).text().trim();

        // Meta from configured selector — "관리자 | 2026-03-09 | 조회수 : 147"
        const metaText = $tr.find(config.selectors.meta).text().trim();
        const parts = metaText.split('|').map((s) => s.trim());
        const author = parts[0] || '';
        const viewsMatch = metaText.match(/조회수\s*:\s*(\d+)/);
        const views = viewsMatch ? parseInt(viewsMatch[1], 10) : 0;

        items.push({
          articleNo,
          title,
          category: '',
          author,
          date,
          views,
          detailPath: href,
        });
      } catch (err) {
        logger.warn({ error: (err as Error).message }, 'Failed to parse gnuboard-custom list item');
      }
    });

    logger.info({ page, count: items.length }, 'Parsed gnuboard-custom list page');
    return items;
  }

  async crawlDetail(ref: DetailRef, config: GnuboardCustomDepartmentConfig): Promise<NoticeDetail | null> {
    let url: string;
    if (ref.detailPath.startsWith('http')) {
      url = ref.detailPath;
    } else if (ref.detailPath.startsWith('?')) {
      url = `${config.baseUrl}${ref.detailPath}`;
    } else {
      url = `${config.baseUrl}?${config.boardParam}=${config.boardName}&mode=${config.detailMode}&num=${ref.articleNo}`;
    }

    try {
      logger.debug({ url, articleNo: ref.articleNo }, 'Fetching gnuboard-custom detail page');
      const html = await this.fetcher.fetch(url);
      const $ = loadHtml(html);

      // Content — clean HWP editor artifacts
      const $content = $(config.selectors.detailContent);
      const rawHtml = $content.html()?.trim() || '';
      const content = cleanHwpArtifacts(rawHtml);
      // Re-parse cleaned content for text extraction
      const $cleaned = loadHtml(content);
      const contentText = $cleaned.text().trim();

      // Attachments: a[href*="download.php"]
      const attachments: { name: string; url: string }[] = [];
      const origin = new URL(config.baseUrl).origin;

      $(config.selectors.detailAttachment).each((_i, el) => {
        const $a = $(el);
        const name = $a.text().trim();
        let fileHref = $a.attr('href') || '';

        if (!name || !fileHref) return;

        // Resolve relative download URLs
        if (fileHref.startsWith('/')) {
          fileHref = `${origin}${fileHref}`;
        } else if (!fileHref.startsWith('http')) {
          fileHref = `${origin}/${fileHref}`;
        }

        attachments.push({ name, url: fileHref });
      });

      return { content, contentText, attachments };
    } catch (err) {
      logger.error({ articleNo: ref.articleNo, error: (err as Error).message }, 'Failed to fetch gnuboard-custom detail');
      return null;
    }
  }
}
